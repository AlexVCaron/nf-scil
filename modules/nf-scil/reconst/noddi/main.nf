
process BETCROP_FSLBETCROP {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_1.6.0.sif':
        'scilus/scilus:1.6.0' }"

    input:
        tuple val(meta), path(dwi), path(bval), path(bvec), path(mask)
        path(kernels)

    output:
        tuple val(meta), path("*__FIT_dir.nii.gz")              , emit: dir
        tuple val(meta), path("*__FIT_ISOVF.nii.gz")            , emit: isovf
        tuple val(meta), path("*__FIT_ICVF.nii.gz")             , emit: icvf
        tuple val(meta), path("*__FIT_OD.nii.gz")               , emit: od
        path "versions.yml"                                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def para_diff = task.ext.para_diff ? "--para_diff " + task.ext.para_diff : ""
    def iso_diff = task.ext.iso_diff ? "--iso_diff " + task.ext.iso_diff : ""
    def lambda1 = task.ext.lambda1 ? "--lambda1 " + task.ext.lambda1 : ""
    def lambda2 = task.ext.lambda2 ? "--lambda2 " + task.ext.lambda2 : ""
    def nb_threads = task.ext.nb_threads ? "--processes " + task.ext.nb_threads : ""
    def b_thr = task.ext.b_thr ? "--b_thr " + task.ext.b_thr : ""

    def set_kernels = []
    if (kernels)
        set_kernels += ["--load_kernels + $kernels"]
    else
        set_kernels += ["--save_kernels kernels/ --compute_only":]

    def set_mask = []
    if (mask) set_mask += ["--mask $mask"]

    """
    scil_compute_NODDI.py $dwi $bval $bvec $set_mask\
        $para_diff $iso_diff $lambda1 $lambda2 $nb_threads $b_thr\
        $set_kernels

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_compute_NODDI.py -h

    touch "${prefix}__FIT_dir.nii.gz"
    touch "${prefix}__FIT_ISOVF.nii.gz"
    touch "${prefix}__FIT_ICVF.nii.gz"
    touch "${prefix}__FIT_OD.nii.gz"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 1.6.0
    END_VERSIONS
    """
}
