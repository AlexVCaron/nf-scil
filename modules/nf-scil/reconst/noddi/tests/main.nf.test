nextflow_process {

    name "Test Process RECONST_NODDI"
    script "../main.nf"
    process "RECONST_NODDI"

    tag "modules"
    tag "modules_nfcore"
    tag "reconst"
    tag "reconst/noddi"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    config "./nextflow.config"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( ["commit_amico.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("reconst - noddi - save_kernels") {
        when {
            process {
            """
                input[0] = LOAD_DATA.out.test_data_directory
                        .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec"),
                        file("\${test_data_directory}/mask.nii.gz"),
                        []
                ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("reconst - noddi - no mask") {
        when {
            process {
            """
                input[0] = LOAD_DATA.out.test_data_directory
                        .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec"),
                        [],
                        []
                ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
