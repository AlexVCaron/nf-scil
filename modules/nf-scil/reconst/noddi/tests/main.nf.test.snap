{
    "reconst - noddi - extract_metrics": {
        "content": [
            [
                "versions.yml:md5,7d60348ede01d2f9bd3a2fe9da61741d"
            ],
            145
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-21T17:05:10.907308"
    },
    "reconst - noddi - save_kernels - no mask": {
        "content": [
            [
                "versions.yml:md5,7d60348ede01d2f9bd3a2fe9da61741d"
            ],
            145
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-21T17:04:36.944861"
    }
}